src/sql/src/plan/query.rs

```diff
@@ -913,6 +913,59 @@ pub fn plan_default_expr(
     Ok((hir, qcx.ids.into_iter().collect()))
 }
 
+pub fn plan_params<'a>(
+    scx: &'a StatementContext,
+    params: Vec<Expr<Raw>>,
+    desc: &StatementDesc,
+) -> Result<Params, anyhow::Error> {
+    if params.len() != desc.param_types.len() {
+        bail!(
+            "expected {} params, got {}",
+            desc.param_types.len(),
+            params.len()
+        );
+    }
+
+    let mut qcx = QueryContext::root(scx, QueryLifetime::OneShot(scx.pcx()?));
+    let scope = Scope::empty(None);
+    let rel_type = RelationType::empty();
+
+    let mut datums = Row::with_capacity(desc.param_types.len());
+    let temp_storage = &RowArena::new();
+    for (mut param, ty) in params.into_iter().zip(&desc.param_types) {
+        transform_ast::transform_expr(scx, &mut param)?;
+        let expr = resolve_names_expr(&mut qcx, param)?;
+
+        let ecx = &ExprContext {
+            qcx: &qcx,
+            name: "EXECUTE",
+            scope: &scope,
+            // TODO(mjibson): Is this correct?
+            relation_type: &rel_type,
+            allow_aggregates: false,
+            allow_subqueries: false,
+        };
+        let ex = plan_expr(ecx, &expr)?
+            .cast_to(
+                ecx.name,
+                ecx,
+                CastContext::Assignment,
+                &ty.to_scalar_type_lossy(),
+            )?
+            .lower_uncorrelated()?;
+        let evaled = ex.eval(&[], temp_storage)?;
+        datums.push(evaled);
+    }
+    Ok(Params {
+        datums,
+        types: desc
+            .param_types
+            .iter()
+            .map(|t| t.to_scalar_type_lossy())
```

Comment:
This looks sketchy though! (The call on L953 also looks sketchy.) Quite possibly it'll mishandle numeric scales or chars or something.

Going from a `pgrepr::Type` to a `ScalarType` is lossy, but going from a `ScalarType` to a `pgrepr::Type` is not. So I think instead you might want to `plan_expr` _without_ attempting to cast to anything. You probably want `.type_as_any()`. And _then_ you probably want to take the type of that expr, convert it to a `pgrepr::Type` and make sure it matches what you expect. Some very rough psuedocode that might be clearer than that verbiage:

```rust
let expr = plan_expr(ecx, &expr).type_as_any(ecx)?;
if pgrepr::Type::from(ecx.scalar_type(expr)) != ty {
    bail!("mismatched blah blah");
}
```

Reply:
TIL! By the the hardest part of this PR was dealing with exprs and scalar/pgrepr types. Thanks for explaining the correct way.